pkg/utils/nodeclaim/nodeclaim.go

```diff
@@ -232,3 +205,39 @@ func UpdateNodeOwnerReferences(nodeClaim *v1beta1.NodeClaim, node *v1.Node) *v1.
 	})
 	return node
 }
+
+// TerminationHandler is a helper function that takes a v1beta1.NodeClaim and calls cloudProvider.Delete() if status condition
+// on nodeClaim is not terminating. If it is terminating then it will check if the kubelet process is still running. It will
+// return an error and a boolean that indicates if the reconciliation should be re-enqueued.
+func TerminationHandler(ctx context.Context, c client.Client, nodeClaim *v1beta1.NodeClaim, node *v1.Node, cloudProvider cloudprovider.CloudProvider) (bool, error) {
+	// Check if the status condition on nodeClaim is Terminating
+	terminationStatus := nodeClaim.StatusConditions().GetCondition(v1beta1.Terminating)
+	if terminationStatus.IsUnknown() || terminationStatus.IsFalse() {
+		// If not then call Delete on cloudProvider to trigger termination and always requeue reconciliation
+		if err := cloudProvider.Delete(ctx, nodeClaim); cloudprovider.IgnoreNodeClaimNotFoundError(err) != nil {
+			return true, fmt.Errorf("terminating cloudprovider instance, %w", err)
+		}
+		return true, SetTerminatingStatus(ctx, c, nodeClaim)
+	}
+	// We do a check on the Ready condition of the node to see if the kubelet process is still running
+	return nodeutils.GetCondition(node, v1.NodeReady).Status == v1.ConditionTrue, nil
```

Comment:
I don't think that we should just signal on the lack of the true condition. I think we should actually be signaling on the Unknown condition. We know that kubelet will post Unknown when it hasn't heard from the process in 40s (see https://kubernetes.io/docs/reference/node/node-status/#:~:text=Unknown%20if%20the%20node%20controller%20has%20not%20heard%20from%20the%20node%20in%20the%20last%20node%2Dmonitor%2Dgrace%2Dperiod), which is an indication that the process is no longer running and we can trust that the node is safe to terminate.